File: lib/agents/competitive.ts
import { searchWeb, searchNews } from '../tools/serpapi';
import { scrapePage, scrapeCompetitorPricing } from '../tools/firecrawl';
import { searchHN } from '../tools/hn-algolia';
import { scrapeTwitterX } from '../tools/apify-twitter';
import { discoverAndScrape } from '../tools/discover-and-scrape';
import { planQueries } from '../tools/query-planner';
import { filterRelevant, requirementsFromContext } from '../tools/relevance';
import { extractCandidates, verifyCandidates } from '../tools/candidate-discovery';
import { generateHuggingFaceJson } from './gemini';
import {
  competitorSiteUrl,
  isPlaceholderCompetitor,
  isUsableScrapePage,
  skippedScrapePromise,
} from './entity-url';
import { runResearchLoop } from './research-loop';
import { applyInsufficientGate, evidencePromptRules } from './evidence-gate';
import { localeFromGeography } from './search-locale';
import {
  resolveEntityProbes,
  resolveGapQueries,
  resolveSearchQueries,
} from './plan-queries';
import type {
  AgentConfig,
  AgentContext,
  AgentOutput,
  CompetitiveOutput,
  CompetitorFeature,
  AgentSource,
  ConfidenceLevel,
  EvidenceCandidate,
} from './types';
import { scoreToLevel } from './types';
import { computeSignalQualityPenalty, extractToolResults } from '../tools/fallback';
import type { SearchResult, ToolResult } from '../tools/types';

async function run(ctx: AgentContext): Promise<AgentOutput> {
  const { query, product, competitor, priorContext, geography, category, namedEntities, requiredTerms } = ctx;

  const competitorName =
    competitor && !isPlaceholderCompetitor(competitor) ? competitor : 'discovered competitors';
  const compUrl = competitorSiteUrl(ctx);
  const locale = localeFromGeography(geography);
  const requirements = requirementsFromContext(ctx);

  const queryBundle = planQueries({
    product,
    competitor,
    domain: 'competitive',
    query,
    category: category ?? (query.toLowerCase().includes('agri') ? 'agritech' : undefined),
    geography,
    namedEntities,
    requiredTerms,
  });

  const planned = resolveSearchQueries(ctx, [queryBundle.broad, queryBundle.targeted], 2);
  const primaryQuery = planned[0] ?? queryBundle.broad;
  const secondaryQuery = planned[1] ?? queryBundle.hypothesis;
  const entityProbes = resolveEntityProbes(ctx, queryBundle.entityProbes, 2);
  const gapQueries = resolveGapQueries(ctx, 3);

  const searchedForBase = [
    primaryQuery,
    secondaryQuery,
    queryBundle.hypothesis,
    ...entityProbes,
    ...gapQueries,
  ];

  let candidates: EvidenceCandidate[] = [];
  let droppedIrrelevantCount = 0;

  const loop = await runResearchLoop({
    domain: 'competitive',
    requirements,
    budgetMs: 15_000,
    round1: () =>
      Promise.allSettled([
        searchWeb(primaryQuery, locale),
        searchNews(queryBundle.hypothesis, locale),
        searchHN(`${product} ${geography?.name ?? ''}`.trim()),
        compUrl ? scrapePage(compUrl) : skippedScrapePromise(),
        compUrl ? scrapeCompetitorPricing(compUrl) : skippedScrapePromise(),
        searchWeb(secondaryQuery, locale),
        scrapeTwitterX(
          [primaryQuery, secondaryQuery],
          { maxItems: 40, sort: 'Latest', language: locale?.hl ?? 'en' },
        ),
        discoverAndScrape(primaryQuery, {
          product,
          competitor: competitorName,
          domain: 'competitive',
          topN: 2,
          keywords: queryBundle.keywords,
          locale,
          requirements,
        }),
        ...entityProbes.map(q => searchWeb(q, locale)),
      ]),
    ingest: (settled, round) => {
      const sources: AgentSource[] = [];
      const rawContent: string[] = [];
      const allHits: SearchResult[] = [];
      let searchCalls = 0;
      let scrapeCalls = 0;
      let dropped = 0;

      const pushHits = (
        result: PromiseSettledResult<unknown>,
        label: string,
        tool: AgentSource['tool'],
        limit = 5,
      ) => {
        if (result.status !== 'fulfilled') return;
        const value = result.value as ToolResult<SearchResult[]>;
        if (!value?.data || !Array.isArray(value.data)) return;
        searchCalls += 1;
        const filtered = filterRelevant(value.data, requirements, { limit, minScore: 0.25 });
        dropped += filtered.dropped.length;
        filtered.kept.forEach(r => {
          allHits.push(r);
          sources.push({ url: r.url, title: r.title, timestamp: value.timestamp, tool });
          rawContent.push(`[${label}] ${r.title}: ${r.snippet}`);
        });
      };

      pushHits(settled[0], 'COMPETITOR WEB', 'serpapi', 5);
      pushHits(settled[1], 'COMPETITOR NEWS', 'serpapi', 4);
      if (settled[2]?.status === 'fulfilled') {
        const hn = settled[2].value as ToolResult<Array<{ url: string; title: string; created: string }>>;
        (hn.data ?? []).slice(0, 3).forEach(p => {
          sources.push({ url: p.url, title: p.title, timestamp: p.created, tool: 'hn' });
          rawContent.push(`[HN] ${p.title}`);
        });
      }
      if (settled[3] && isUsableScrapePage(settled[3] as PromiseSettledResult<ToolResult<import('../tools/types').ScrapedPage>>)) {
        scrapeCalls += 1;
        const page = (settled[3] as PromiseFulfilledResult<ToolResult<import('../tools/types').ScrapedPage>>).value.data;
        sources.push({ url: page.url, title: page.title || competitorName, timestamp: (settled[3] as PromiseFulfilledResult<ToolResult<import('../tools/types').ScrapedPage>>).value.timestamp, tool: 'firecrawl' });
        rawContent.push(`[COMPETITOR HOMEPAGE] ${page.excerpt}`);
      }
      if (settled[4] && isUsableScrapePage(settled[4] as PromiseSettledResult<ToolResult<import('../tools/types').ScrapedPage>>)) {
        scrapeCalls += 1;
        const page = (settled[4] as PromiseFulfilledResult<ToolResult<import('../tools/types').ScrapedPage>>).value.data;
        sources.push({ url: page.url, title: `${competitorName} pricing`, timestamp: (settled[4] as PromiseFulfilledResult<ToolResult<import('../tools/types').ScrapedPage>>).value.timestamp, tool: 'firecrawl' });
        rawContent.push(`[COMPETITOR PRICING] ${page.excerpt}`);
      }
      pushHits(settled[5], 'SOCIAL SIGNAL', 'serpapi', 3);
      if (settled[6]?.status === 'fulfilled') {
        const tw = settled[6].value as ToolResult<Array<{ url: string; authorHandle?: string; text: string; createdAt?: string }>>;
        (tw.data ?? []).slice(0, 6).forEach(t => {
          sources.push({
            url: t.url,
            title: `X @${t.authorHandle ?? 'unknown'}`,
            timestamp: t.createdAt ?? tw.timestamp,
            tool: 'apify',
          });
          rawContent.push(`[APIFY X] @${t.authorHandle ?? 'unknown'}: ${t.text}`);
        });
      }
      if (settled[7]?.status === 'fulfilled') {
        const disc = settled[7].value as Awaited<ReturnType<typeof discoverAndScrape>>;
        dropped += disc.droppedIrrelevantCount ?? 0;
        searchCalls += 1;
        disc.pages.forEach(pageResult => {
          if (pageResult.status === 'failed' || !pageResult.data.markdown?.trim()) return;
          scrapeCalls += 1;
          const page = pageResult.data;
          sources.push({
            url: page.url,
            title: page.title || competitorName,
            timestamp: pageResult.timestamp,
            tool: 'firecrawl',
          });
          rawContent.push(`[DISCOVERED PAGE] ${page.title}: ${page.excerpt}`);
        });
        filterRelevant(disc.search.data, requirements, { limit: 5 }).kept.forEach(r => {
          allHits.push(r);
        });
      }
      for (let i = 8; i < settled.length; i++) {
        pushHits(settled[i], 'ENTITY PROBE', 'serpapi', 4);
      }

      if (round === 2) {
        for (const s of settled) {
          if (s.status !== 'fulfilled') continue;
          const val = s.value as Awaited<ReturnType<typeof discoverAndScrape>> | ToolResult<SearchResult[]>;
          if (val && 'pages' in val && Array.isArray(val.pages)) {
            val.pages.forEach(pageResult => {
              if (pageResult.status === 'failed' || !pageResult.data.markdown?.trim()) return;
              scrapeCalls += 1;
              const page = pageResult.data;
              sources.push({
                url: page.url,
                title: page.title || 'Candidate site',
                timestamp: pageResult.timestamp,
                tool: 'firecrawl',
              });
              rawContent.push(`[CANDIDATE SITE] ${page.title}: ${page.excerpt}`);
            });
          } else if (val && 'data' in val && Array.isArray(val.data)) {
            pushHits(
              { status: 'fulfilled', value: val } as PromiseFulfilledResult<ToolResult<SearchResult[]>>,
              'GAP QUERY',
              'serpapi',
              4,
            );
          }
        }
      }

      if (round === 1 && allHits.length > 0) {
        candidates = extractCandidates(allHits, {
          geographyName: geography?.name,
          exclude: [product, competitorName, ...(namedEntities ?? [])],
          limit: 5,
        });
      }

      droppedIrrelevantCount += dropped;

      return {
        sources,
        rawContent,
        toolResults: extractToolResults(settled as PromiseSettledResult<ToolResult<unknown>>[]),
        searchedFor: searchedForBase,
        relevantSourceCount: sources.length,
        relevantHits: allHits,
        scrapedPageCount: sources.filter(s => s.tool === 'firecrawl').length,
        droppedIrrelevantCount: dropped,
        searchCallCount: searchCalls,
        scrapeCallCount: scrapeCalls,
      };
    },
    gapRound: async (state) => {
      if (gapQueries.length > 0) {
        const settled = await Promise.allSettled(
          gapQueries.slice(0, 2).map(q =>
            discoverAndScrape(q, {
              product,
              competitor: competitorName,
              domain: 'competitive',
              topN: 1,
              keywords: queryBundle.keywords,
              locale,
              requirements,
            }),
          ),
        );
        searchedForBase.push(...gapQueries.slice(0, 2));
        return settled;
      }

      if (candidates.length === 0) {
        candidates = extractCandidates(
          state.sources.map(s => ({ title: s.title, url: s.url, snippet: '' })),
          { geographyName: geography?.name, exclude: [product], limit: 5 },
        );
      }
      const { settled, queries } = await verifyCandidates(candidates, {
        product,
        geographyName: geography?.name,
        category: category ?? queryBundle.keywords[0],
        maxCandidates: 3,
        topN: 1,
      });
      searchedForBase.push(...queries);
      return settled;
    },
  });

  if (candidates.length > 0) {
    loop.evidence = { ...loop.evidence, candidates };
  }

  const systemPrompt = `You are a competitive intelligence analyst. You compare product capabilities with brutal honesty. You separate facts from interpretation. You never fabricate features.
${priorContext ? `\nPrior conversation context:\n${priorContext}` : ''}
${evidencePromptRules(geography, category)}`;

  const userPrompt = `Query: "${query}"
Our product: ${product}
Competitor focus: ${competitorName}
${geography ? `Geography constraint: ${geography.name}` : ''}
${category ? `Category: ${category}` : ''}
Evidence status: ${loop.evidence.status} (${loop.evidence.relevantSourceCount} relevant sources)
${candidates.length ? `Candidate organisations found: ${candidates.map(c => `${c.name} (${c.classification})`).join(', ')}` : ''}
${ctx.discoveredEntities?.length ? `Planner entities: ${ctx.discoveredEntities.map(e => e.name).join(', ')}` : ''}

Raw signals:
${(await import('@/lib/guardrails')).fenceUntrusted(loop.rawContent)}

Produce a JSON object:
{
  "insufficientEvidence": boolean,
  "facts": string[],
  "interpretation": string[],
  "competitorSummary": string,
  "matrix": [
    {
      "feature": string,
      "yourProduct": "strong" | "medium" | "weak" | "none",
      "competitor": "strong" | "medium" | "weak" | "none",
      "gapDirection": "advantage" | "parity" | "disadvantage"
    }
  ],
  "hiringSignals": string[],
  "recentMoves": string[],
  "synthesizedAnswer": string,
  "confidenceScore": number
}

For the matrix, only include dimensions backed by signals. If insufficientEvidence, leave matrix empty.`;

  let parsed: Record<string, unknown> = {};
  try {
    parsed = await generateHuggingFaceJson<Record<string, unknown>>(systemPrompt, userPrompt, {
      maxNewTokens: 1400,
      temperature: 0.2,
    });
  } catch {
    parsed = {
      insufficientEvidence: loop.evidence.status === 'insufficient',
      facts: loop.rawContent.slice(0, 3).map(s => s.replace(/^\[[^\]]+\]\s*/, '')).filter(s => s.length > 15),
      interpretation: ['Analysis synthesis is temporarily unavailable.'],
      competitorSummary: '',
      matrix: [],
      hiringSignals: [],
      recentMoves: [],
      synthesizedAnswer: '',
      confidenceScore: 0.4,
    };
  }

  const gate = applyInsufficientGate(parsed, loop.evidence);
  if (gate.insufficient) {
    loop.evidence = { ...loop.evidence, status: 'insufficient' };
  }

  const confScore = Number.parseFloat(
    (gate.confidenceScore * computeSignalQualityPenalty(loop.toolResults, 8)).toFixed(2),
  );
  const confidence: ConfidenceLevel = scoreToLevel(confScore);

  const output: CompetitiveOutput = {
    agentId: 'competitive',
    domain: 'competitive',
    artifactType: 'competitive-matrix',
    confidence,
    confidenceScore: confScore,
    facts: gate.insufficient ? (parsed.facts as string[] ?? []).slice(0, 2) : (parsed.facts as string[] ?? []),
    interpretation: gate.insufficient ? loop.evidence.gaps : (parsed.interpretation as string[] ?? []),
    sources: loop.sources,
    generatedAt: new Date().toISOString(),
    competitor: competitorName,
    matrix: gate.insufficient ? [] : ((parsed.matrix ?? []) as CompetitorFeature[]),
    competitorSummary: gate.insufficient
      ? loop.evidence.gaps[0] ?? 'Insufficient competitive evidence for this market.'
      : (parsed.competitorSummary as string) ?? '',
    hiringSignals: gate.insufficient ? [] : (parsed.hiringSignals as string[] ?? []),
    recentMoves: gate.insufficient ? [] : (parsed.recentMoves as string[] ?? []),
    evidence: loop.evidence,
    toolCallCount: loop.toolCallCount,
    searchCallCount: loop.searchCallCount,
    scrapeCallCount: loop.scrapeCallCount,
    droppedIrrelevantCount: droppedIrrelevantCount + (loop.droppedIrrelevantCount ?? 0),
  };

  return output;
}

export const competitiveAgent: AgentConfig = {
  id: 'competitive',
  name: 'Competitive Agent',
  description: 'Scrapes competitor product pages, changelogs, and pricing to build a feature comparison matrix.',
  run,
};

File: backend/app/agents/competitive.py
"""Competitive Landscape agent (parity with lib/agents/competitive.ts)."""

from __future__ import annotations

import asyncio
import re
from datetime import datetime, timezone
from typing import Any

from app.agents.entity_url import (
    competitor_site_url,
    is_usable_scrape_page,
    skipped_scrape_promise,
)
from app.gemini import generate_json
from app.models import score_to_level
from app.tools.apify_twitter import scrape_twitter_x
from app.tools.fallback import compute_signal_quality_penalty, extract_tool_results
from app.tools.firecrawl import scrape_competitor_pricing, scrape_page
from app.tools.hn_algolia import search_hn
from app.tools.serpapi import search_news, search_web

AGENT_ID = "competitive"
AGENT_NAME = "Competitive Agent"
AGENT_DESCRIPTION = (
    "Scrapes competitor product pages, changelogs, and pricing to build a feature comparison matrix."
)


def _fulfilled(result: Any) -> bool:
    return not isinstance(result, BaseException)


async def run(ctx: dict[str, Any]) -> dict[str, Any]:
    query = ctx["query"]
    product = ctx["product"]
    competitor = ctx.get("competitor")
    prior_context = ctx.get("prior_context")

    competitor_name = competitor or "relevant competitors"
    comp_url = competitor_site_url(ctx)

    (
        web_result,
        news_result,
        hn_result,
        scrape_result,
        pricing_result,
        social_signals_result,
        apify_twitter_result,
    ) = await asyncio.gather(
        search_web(f"{competitor_name} features product update 2025 2026"),
        search_news(f"{competitor_name} funding launch product announcement 2025"),
        search_hn(f"{competitor_name} {product}"),
        scrape_page(comp_url) if comp_url else skipped_scrape_promise(),
        scrape_competitor_pricing(comp_url) if comp_url else skipped_scrape_promise(),
        search_web(
            f"{competitor_name} {product} "
            "site:x.com OR site:twitter.com OR site:instagram.com OR site:linkedin.com launch feature feedback"
        ),
        scrape_twitter_x(
            [f"{competitor_name} {product}", f"{competitor_name} launch feedback"],
            max_items=80,
            sort="Latest",
            language="en",
        ),
        return_exceptions=True,
    )

    (hiring_result,) = await asyncio.gather(
        search_web(
            f'{competitor_name} jobs hiring "AI" OR "machine learning" OR "sales" '
            "site:linkedin.com OR site:greenhouse.io"
        ),
        return_exceptions=True,
    )

    sources: list[dict[str, Any]] = []
    raw_content: list[str] = []

    if _fulfilled(web_result):
        for r in (web_result.get("data") or [])[:5]:
            sources.append(
                {
                    "url": r["url"],
                    "title": r["title"],
                    "timestamp": web_result["timestamp"],
                    "tool": "serpapi",
                }
            )
            raw_content.append(f"[COMPETITOR WEB] {r['title']}: {r['snippet']}")

    if _fulfilled(news_result):
        for r in (news_result.get("data") or [])[:4]:
            sources.append(
                {
                    "url": r["url"],
                    "title": r["title"],
                    "timestamp": news_result["timestamp"],
                    "tool": "serpapi",
                }
            )
            raw_content.append(f"[COMPETITOR NEWS] {r['title']}: {r['snippet']}")

    if _fulfilled(hn_result):
        for p in (hn_result.get("data") or [])[:3]:
            sources.append(
                {
                    "url": p["url"],
                    "title": p["title"],
                    "timestamp": p["created"],
                    "tool": "hn",
                }
            )
            raw_content.append(f"[HN] {p['title']}")

    if is_usable_scrape_page(scrape_result):
        page = scrape_result["data"]
        sources.append(
            {
                "url": page["url"],
                "title": page.get("title") or competitor_name,
                "timestamp": scrape_result["timestamp"],
                "tool": "firecrawl",
            }
        )
        raw_content.append(f"[COMPETITOR HOMEPAGE] {page['excerpt']}")

    if is_usable_scrape_page(pricing_result):
        page = pricing_result["data"]
        label = f"{competitor} pricing" if competitor else "Competitor pricing page"
        sources.append(
            {
                "url": page["url"],
                "title": label,
                "timestamp": pricing_result["timestamp"],
                "tool": "firecrawl",
            }
        )
        raw_content.append(f"[COMPETITOR PRICING] {page['excerpt']}")

    if _fulfilled(social_signals_result):
        for r in (social_signals_result.get("data") or [])[:3]:
            sources.append(
                {
                    "url": r["url"],
                    "title": r["title"],
                    "timestamp": social_signals_result["timestamp"],
                    "tool": "serpapi",
                }
            )
            raw_content.append(f"[SOCIAL SIGNAL] {r['title']}: {r['snippet']}")

    if _fulfilled(apify_twitter_result):
        for t in (apify_twitter_result.get("data") or [])[:8]:
            handle = t.get("authorHandle") or t.get("author_handle") or "unknown"
            sources.append(
                {
                    "url": t["url"],
                    "title": f"X @{handle}",
                    "timestamp": t.get("createdAt") or t.get("created_at") or apify_twitter_result["timestamp"],
                    "tool": "apify",
                }
            )
            likes = t.get("likeCount") if isinstance(t.get("likeCount"), (int, float)) else t.get("like_count")
            likes_suffix = f" (likes {likes})" if isinstance(likes, (int, float)) else ""
            raw_content.append(f"[APIFY X] @{handle}: {t['text']}{likes_suffix}")

    if _fulfilled(hiring_result):
        for r in (hiring_result.get("data") or [])[:3]:
            raw_content.append(f"[HIRING SIGNAL] {r['title']}: {r['snippet']}")

    prior_block = f"\nPrior conversation context:\n{prior_context}" if prior_context else ""
    system_prompt = f"""You are a competitive intelligence analyst specialising in B2B SaaS. You compare product capabilities with brutal honesty. You separate facts from interpretation. You never fabricate features.{prior_block}"""

    user_prompt = f"""Query: "{query}"
Our product: {product}
Competitor: {competitor_name}

Raw signals:
{chr(10).join(raw_content)}

Produce a JSON object:
{{
  "facts": string[],
  "interpretation": string[],
  "competitorSummary": string,
  "matrix": [
    {{
      "feature": string,
      "yourProduct": "strong" | "medium" | "weak" | "none",
      "competitor": "strong" | "medium" | "weak" | "none",
      "gapDirection": "advantage" | "parity" | "disadvantage"
    }}
  ],
  "hiringSignals": string[],
  "recentMoves": string[],
  "synthesizedAnswer": string,
  "confidenceScore": number
}}

For the matrix, infer the most relevant feature dimensions from the signals above. Choose dimensions that are actually relevant to {product} and {competitor_name} based on what the data shows."""

    try:
        parsed = await generate_json(system_prompt, user_prompt, max_new_tokens=1400, temperature=0.2)
    except Exception:
        parsed = {
            "facts": [
                re.sub(r"^\[[^\]]+\]\s*", "", s)
                for s in raw_content[:3]
                if len(re.sub(r"^\[[^\]]+\]\s*", "", s)) > 15
            ],
            "interpretation": [
                "Analysis synthesis is temporarily unavailable. Raw data signals are shown below."
            ],
            "competitorSummary": f"{competitor_name} competitive data collected.",
            "matrix": [],
            "hiringSignals": [],
            "recentMoves": [],
            "synthesizedAnswer": "Competitive data was gathered but synthesis failed.",
            "confidenceScore": 0.4,
        }

    raw_score = parsed["confidenceScore"] if isinstance(parsed.get("confidenceScore"), (int, float)) else 0.6
    tool_results = extract_tool_results(
        [
            web_result,
            news_result,
            hn_result,
            scrape_result,
            pricing_result,
            social_signals_result,
            apify_twitter_result,
            hiring_result,
        ]
    )
    conf_score = round(float(raw_score) * compute_signal_quality_penalty(tool_results, 8), 2)
    confidence = score_to_level(conf_score)

    return {
        "agentId": AGENT_ID,
        "domain": "competitive",
        "artifactType": "competitive-matrix",
        "confidence": confidence,
        "confidenceScore": conf_score,
        "facts": parsed.get("facts") or [],
        "interpretation": parsed.get("interpretation") or [],
        "sources": sources,
        "generatedAt": datetime.now(timezone.utc).isoformat(),
        "competitor": competitor_name,
        "matrix": parsed.get("matrix") or [],
        "competitorSummary": parsed.get("competitorSummary") or "",
        "hiringSignals": parsed.get("hiringSignals") or [],
        "recentMoves": parsed.get("recentMoves") or [],
    }
